### PR TITLE
to_labelled() should identify factors coded as "[code] label"

### LINKED ISSUE #74: Account for prefixed factors to define code-label association automatically in to_labelled.factor
Hi @larmarange,

Sometime ago I had started to work on a function that allows to return to labelled vectors from factors which I called `as_labelled`.

I have encountered several use cases in which it was required in a marketing survey company (for example because some users where working in R and others in SPSS, with constant back and forth).

The idea was to exploit the behavior of `haven::as_factor(x, levels = "both")` when creating factors as it keep both the information of the values and labels sticking to a consistent format.

Here is a simple example:
```r
foo_lbl <- haven::labelled(c(1, 2, 3, 2, 1),
                           labels = c("one" = 1, "two" = 2, "three" = 3),
                           label = "test")
foo_lbl
#> <labelled<double>[5]>: test
#> [1] 1 2 3 2 1
#> 
#> Labels:
#>  value label
#>      1   one
#>      2   two
#>      3 three

foo_fct <- haven::as_factor(foo_lbl, levels = "both")
foo_fct
#> [1] [1] one   [2] two   [3] three [2] two   [1] one  
#> attr(,"label")
#> [1] test
#> Levels: [1] one [2] two [3] three

foo_fct_labelled <- as_labelled(foo_fct)
foo_fct_labelled
#> <labelled<double>[5]>: test
#> [1] 1 2 3 2 1
#> 
#> Labels:
#>  value label
#>      1   one
#>      2   two
#>      3 three

identical(foo_lbl, foo_fct_labelled)
#> [1] TRUE
```

Here is what my old code looks like:

``` r
as_labelled <- function(x, label = NULL) {
  ## handling some wrong inputs:
  if (any(class(x) %in% "haven_labelled")) {
    return(x)
  }

  if (!any(class(x) %in% c("character", "factor"))) {
    stop("This function only handles variables of type character or factor",
         call. = FALSE)
  }
  
  ## extracting values and labels:
  matrix_matching <- stringr::str_match(x, "^\\[(.+)\\]\\s(.+)$")
  tbl_matching <- tibble::as_tibble(matrix_matching, 
                                    .name_repair = ~ make.names(names = c("string", "value", "description")))

  ## converting values to numeric:
  tbl_matching$value <- as.numeric(tbl_matching$value)

  ## check that the input format is correct (after trimming out the cases for which values are NA):
  if (any(is.na(tbl_matching[!is.na(x), ]))) {
    stop("This function can only handle levels defined as '[x] y' with 'x' being a number and 'y' some text",
         call. = FALSE)
  }

  ## preparing labels:
  distinct_labels <- na.omit(dplyr::distinct(tbl_matching[, c("value", "description")]))
  distinct_labels <- dplyr::arrange(distinct_labels, value)

  labels <- as.numeric(distinct_labels$value)
  names(labels) <- distinct_labels$description

  ## create the variable label:
  if (!is.null(label)) {
    var_label <- label
  } else {
    var_label <- attributes(x)[["label"]]
  }

  ## create and return labelled variable:
  haven::labelled(tbl_matching$value, labels = labels, label = var_label)
}
```

It could certainly been improved and should be thoroughly tested in corner cases.
The dependencies to {stringr} can probably been removed without much difficulty by grepping.
It could also benefit in relying on the existing functions of your package.

If it would be absorbed somehow in your package, it could help me to distribute this solution and perhaps you find the whole thing not completely useless.

Please let me know if you are interested, and I would be happy to draft a proper pull request if you do (not immediately but within the next couple of weeks).

Best,
Alex

### PR DESCRIPTION
**Pull request by larmarange**
fix #74

### CODE REVIEW

**Comment by larmarange** [comment]
``` r
library(labelled)
l <- labelled(c(1, 1, 2, 2, 9, 2, 1, 9), c("yes" = 1, "no" = 2, "don't know" = 9))
f <- to_factor(l, levels = "p")
f
#> [1] [1] yes        [1] yes        [2] no         [2] no         [9] don't know
#> [6] [2] no         [1] yes        [9] don't know
#> Levels: [1] yes [2] no [9] don't know
to_labelled(f)
#> <labelled<double>[8]>
#> [1] 1 1 2 2 9 2 1 9
#> 
#> Labels:
#>  value      label
#>      1        yes
#>      2         no
#>      9 don't know
identical(to_labelled(f), l)
#> [1] TRUE
```

<sup>Created on 2020-11-16 by the [reprex package](https://reprex.tidyverse.org) (v0.3.0)</sup>

File: R/to_labelled.R [hunk 1]
```diff
@@ -206,21 +210,43 @@ memisc_to_labelled <- function(x) {
...
+      names(labs) <- levels(x)
+      x <- labelled(as.numeric(x), labs)
+    } else {
+      # "[code] label" case
+      if (all(!is.na(suppressWarnings(as.numeric(l$code)))))
+        l$code <- as.numeric(l$code)
+        r <- l$levels
+        names(r) <- l$code
+        x <- forcats::fct_recode(x, !!!r) %>%
+          as.character()
```
Comment (courtiol) [inline R/to_labelled.R:236]:
```suggestion
        levels(x) <- l$code
        x <- as.character(x)
```
That would avoid the new dependency to forcats

File: DESCRIPTION [hunk 2]
```diff
@@ -19,6 +19,7 @@ Encoding: UTF-8
 Imports:
     haven (>= 2.3.1),
     dplyr,
+    forcats,
```
Comment (courtiol) [inline DESCRIPTION:22]:
```suggestion
```
Perhaps not needed, see below

File: R/is_prefixed.R [hunk 3]
```diff
@@ -0,0 +1,22 @@
+#' Check if a factor is prefixed
+#' @param x a factor
+#' @export
+is_prefixed <- function(x) {
+  if (!is.factor(x))
+    stop("is_prefixed should be used only with a factor.")
+  l <- .get_prefixes.factor(x)
+  all(!is.na(l$code)) & all(!is.na(l$code))
```
Comment (courtiol) [inline R/is_prefixed.R:8]:
```suggestion
  all(!is.na(l$code)) && all(!is.na(l$code))
```
If both are scalar, it could save a nanosecond ;-)

**Review by courtiol — APPROVED** [review]
Minor suggestions, including one not to increase dependencies.

File: R/to_labelled.R (hunk 1 above)
Comment (larmarange) [reply R/to_labelled.R:236]: mais bien sûr !

simple and elegant. I knew I was missing something here

**Pull request merged**